apis/metal3.io/v1alpha1/baremetalhost_validation.go

```diff
@@ -0,0 +1,55 @@
+package v1alpha1
+
+import (
+	"fmt"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// log is for logging in this package.
+var log = logf.Log.WithName("baremetalhost-validation")
+
+// validateHost validates BareMetalHost resource for creation
+func (host *BareMetalHost) validateHost() []error {
+	log.Info("validate create", "name", host.Name)
+	var errs []error
+
+	if err := validateRAID(host.Spec.RAID); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errs
+}
+
+// validateChanges validates BareMetalHost resource on changes
+// but also covers the validations of creation
+func (host *BareMetalHost) validateChanges(old *BareMetalHost) []error {
+	log.Info("validate update", "name", host.Name)
+	var errs []error
+
+	if err := host.validateHost(); err != nil {
+		errs = append(errs, err...)
+	}
+
+	if old.Spec.BMC.Address != "" && host.Spec.BMC.Address != old.Spec.BMC.Address {
+		errs = append(errs, fmt.Errorf("BMC address can not be changed once it is set"))
+	}
+
+	if old.Spec.BootMACAddress != "" && host.Spec.BootMACAddress != old.Spec.BootMACAddress {
```

Comment:
(not blocking) for the sake of maintenance I'd find useful to have a separate BMH method for each distinct validation rule, ie `validateBMCAddress(), validateBootMacAddress()` and so on